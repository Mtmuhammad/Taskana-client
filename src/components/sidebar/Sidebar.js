// Sidebar navigation 

import "./Sidebar.scss";
import { useNavigate, NavLink } from "react-router-dom";
import { useState, useEffect } from "react";
import useAuth from "../../hooks/useAuth";
import useLogout from "../../hooks/useLogout";

const body = document.querySelector("body");

const Sidebar = ({ page }) => {
  const { auth } = useAuth();
  const [isOpen, setIsOpen] = useState(true);
  const [isDark, setIsDark] = useState(
    localStorage.getItem("isDark") === "true"
  );
  const logout = useLogout()
  const navigate = useNavigate()

  const signOut = async () => {
    await logout()
    navigate('/')
  }

  //save dark/light mode to local storage
  useEffect(() => {
    localStorage.setItem("isDark", JSON.stringify(isDark));
  }, [isDark]);

  // toggle light/dark mode 
  const toggleDark = () => {
    body.classList.toggle("dark");
    body.classList.contains("dark") ? setIsDark(true) : setIsDark(false);
  };

  return (
    <>
      <nav className={isOpen ? "sidebar" : "sidebar close"}>
        <header>
          <div className="image-text">
            <span className="image">
              <img src="taskana-alt.svg" alt="Taskana alt" />
            </span>

            <div className="text logo-text">
              <span data-testid="first-name" className="name">{auth?.user?.firstName}</span>
              <span data-testid="last-name" className="name"> {auth?.user?.lastName}</span>
              <span data-testid="role" className="role">{auth?.user?.empRole}</span>
            </div>
          </div>
          <i
            onClick={() => setIsOpen(!isOpen)}
            className="bx bx-menu toggle"
          ></i>
        </header>

        <div className="menu-bar">
          <div className="menu">
            <li
              onClick={() => !isOpen && setIsOpen(true)}
              className="search-box"
            >
              <i className="bx bx-search icon"></i>
              <input type="text" placeholder="Search..." />
            </li>

            <ul className="p-0 menu-links">
              <li className="nav-link">
                <NavLink
                data-testid="sidebar-link"
                  to="/"
                  className={({ isActive }) =>
                    isActive ? "sidebar-links active" : "sidebar-links"
                  }
                  end
                >
                  <i className="bx bx-home-alt icon"></i>
                  <span className="text nav-text">Dashboard</span>
                </NavLink>
              </li>

              <li className="nav-link">
                <NavLink
                data-testid="sidebar-link"
                  to="/projects"
                  className={({ isActive }) =>
                    isActive ? "sidebar-links active" : "sidebar-links"
                  }
                >
                  <i className="bx bx-briefcase-alt icon"></i>
                  <span className="text nav-text">Projects</span>
                </NavLink>
              </li>

              <li className="nav-link">
                <NavLink
                data-testid="sidebar-link"
                  to="/tasks"
                  className={({ isActive }) =>
                    isActive ? "sidebar-links active" : "sidebar-links"
                  }
                >
                  <i className="bx bx-clipboard icon"></i>
                  <span className="text nav-text">Tasks</span>
                </NavLink>
              </li>

              <li className="nav-link">
                <NavLink
                data-testid="sidebar-link"
                  to="/tickets"
                  className={({ isActive }) =>
                    isActive ? "sidebar-links active" : "sidebar-links"
                  }
                >
                  <i className="bx bx-note icon"></i>
                  <span className="text nav-text">Tickets</span>
                </NavLink>
              </li>

              <li className="nav-link">
                <NavLink
                data-testid="sidebar-link"
                  to="/team"
                  className={({ isActive }) =>
                    isActive ? "sidebar-links active" : "sidebar-links"
                  }
                >
                  <i className="bx bx-group icon"></i>
                  <span className="text nav-text">Team</span>
                </NavLink>
              </li>

              <li className="nav-link">
                <NavLink
                data-testid="sidebar-link"
                  to="/account"
                  className={({ isActive }) =>
                    isActive ? "sidebar-links active" : "sidebar-links"
                  }
                >
                  <i className="bx bx-folder-open icon"></i>
                  <span className="text nav-text">Account</span>
                </NavLink>
              </li>
            </ul>
          </div>

          <div className="bottom-content">
            <li className="">
              <NavLink
              onClick={signOut}
              data-testid="sidebar-link"
                to="/logout"
                className={({ isActive }) =>
                  isActive ? "sidebar-links active" : "sidebar-links"
                }
              >
                <i className="bx bx-log-out icon"></i>
                <span className="text nav-text">Logout</span>
              </NavLink>
            </li>

            <li className="mode">
              <div className="sun-moon">
                <i className="bx bx-moon icon moon"></i>
                <i className="bx bx-sun icon sun"></i>
              </div>
              <span className="mode-text text">
                {isDark ? "Light Mode" : "Dark Mode"}
              </span>

              <div data-testid="toggle-switch" onClick={toggleDark} className="toggle-switch">
                <span className="switch"></span>
              </div>
            </li>
          </div>
        </div>
      </nav>
    </>
  );
};

export default Sidebar;
