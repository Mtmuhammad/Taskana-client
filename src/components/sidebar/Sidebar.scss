@use "../../styles/abstracts/_variables" as *;
@use "../../styles/abstracts/_mixins" as *;

.menu-bar::webkit-scrollbar {
  display: none;
}

.menu-bar .bottom-content .toggle-switch {
  position: absolute;
  right: 0;
  height: 100%;
  min-width: 6rem;
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 0.6rem;
  cursor: pointer;
}
.toggle-switch .switch {
  position: relative;
  height: 2.2rem;
  width: 4rem;
  border-radius: 2.5rem;
  background-color: var(--toggle-color);
  transition: var(--tran-05);
}

.switch::before {
  content: "";
  position: absolute;
  height: 1.5rem;
  width: 1.5rem;
  border-radius: 50%;
  top: 50%;
  left: 0.5rem;
  transform: translateY(-50%);
  background-color: var(--sidebar-color);
  transition: var(--tran-04);
}

body.dark .switch::before {
  left: 2rem;
}

.pages {
  position: absolute;
  top: 0;
  top: 0;
  left: 23.7rem;
  height: 100vh;
  width: calc(100% - 23.7rem);
  background-color: var(--body-color);
  transition: all 0.3s ease;
  padding: 4rem;
}


.sidebar.close ~ .pages {
  left: 7.8rem;
  height: 100vh;
  width: calc(100% - 7.8rem);
}


.mode {
  border-radius: 0.6rem;
  background-color: var(--primary-color-light);
  position: relative;
  transition: all 0.3s ease;

  &-text {
    font-size: 1.5rem!important;
  }
  & .sun-moon {
    height: 5rem;
    width: 6rem;

    & i {
      position: absolute;

      &.sun {
        opacity: 0;
      }
    }
  }
}

.sidebar {
  position: fixed;
  top: 0;
  left: 0;
  height: 100%;
  width: 25rem;
  padding: 1rem 1.4rem;
  background: var(--sidebar-color);
  transition: all 0.3s ease;
  z-index: 100;

  @include respond(sphone) {
    width: 21rem;
  }

  & .menu-bar {
    height: calc(100% - 9.5rem);
    display: flex;
    flex-direction: column;
    justify-content: space-between;
  }

  & .menu {
    margin-top: 2.5rem;
  }

  &.close {
    width: 8.8rem;

    & .toggle {
      transform: translateY(-50%) rotate(0deg);
    }

    & .text {
      opacity: 0;
    }
  }

  & li {
    height: 5rem;
    list-style: none;
    display: flex;
    align-items: center;
    margin-top: 1rem;

    & .sidebar-links {
      list-style: none;
      height: 100%;
      background-color: transparent;
      display: flex;
      align-items: center;
      height: 100%;
      width: 100%;
      border-radius: 0.6rem;
      text-decoration: none;
      transition: all 0.3s ease;

      &.active {
        background-color: $primary-green;

        & .icon,
        & .text {
          color: $white;
        }
      }

      &:hover {
        background-color: var(--primary-color);
      }

      &:hover .icon,
      &:hover .text,
      &.active {
        color: var(--sidebar-color);
      }
    }

    &.search-box {
      border-radius: 0.6rem;
      background-color: var(--primary-color-light);
      cursor: pointer;
      transition: all 0.3s ease;

      & input {
        height: 100%;
        width: 100%;
        outline: none;
        border: none;
        background-color: var(--primary-color-light);
        color: var(--text-color);
        border-radius: 0.6rem;
        font-size: 1.7rem;
        font-weight: 400;
        transition: all 0.3s ease;
      }
    }
  }

  & header .image,
  & .icon {
    min-width: 6rem;
    border-radius: 0.6rem;
  }

  & .icon {
    min-width: 6rem;
    border-radius: 0.6rem;
    height: 100%;
    display: flex;
    align-items: center;
    justify-content: center;
    font-size: 2rem;
  }

  & .text,
  & .icon {
    color: var(--text-color);
    transition: all 0.3s ease;
  }

  & .text {
    font-size: 1.9rem;
    font-weight: 400;
    white-space: nowrap;
    opacity: 1;
  }

  & header {
    position: relative;

    & .image-text {
      display: flex;
      align-items: center;

      & .role {
        font-size: 1.6rem;
        margin-top: -0.2rem;
        display: block;
      }
    }

    & .logo-text {
      display: flex;
      flex-direction: column;
    }

    & .name {
      margin-top: 0.2rem;
      font-size: 1.8rem;
      font-weight: 700;
    }

    & .image {
      display: flex;
      align-items: center;
      justify-content: center;

      & img {
        width: 4rem;
        border-radius: 0.6rem;
      }
    }

    & .toggle {
      position: absolute;
      top: 50%;
      right: -3.5rem;
      transform: translateY(-50%) rotate(180deg);
      height: 4rem;
      width: 4rem;
      background-color: $primary-green;
      color: var(--sidebar-color);
      border-radius: 50%;
      display: flex;
      align-items: center;
      justify-content: center;
      font-size: 3.2rem;
      cursor: pointer;
      transition: all 0.3s ease;
    }
  }
}

body.dark .sidebar header .toggle {
  color: var(--sidebar-color);
}

body.dark .sidebar li a:hover {
  & .icon,
  & .text {
    color: var(--text-color);
  }
}

body.dark .mode .sun-moon i {
  &.sun {
    opacity: 1;
  }

  &.moon {
    opacity: 0;
  }
}
