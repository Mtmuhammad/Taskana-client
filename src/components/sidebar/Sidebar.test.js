import React from "react";
import { render, screen } from "@testing-library/react";
import Sidebar from "./Sidebar";
import {BrowserRouter} from "react-router-dom"


describe("smoke and snapshot tests", () => {
  test("should render without crashing", () => {
    render(<Sidebar/>, {wrapper: BrowserRouter});
  });

  test("should match snapshot", () => {
    const { asFragment } = render(<Sidebar />, {wrapper: BrowserRouter});
    expect(asFragment()).toMatchSnapshot();
  });
});

test("should render nav", () => {
   render(<Sidebar/>, {wrapper: BrowserRouter});
   const nav = screen.getByRole("navigation")
   expect(nav).toBeInTheDocument()
   expect(nav).toHaveClass("sidebar")
})

describe("header", () => {
   test("should render header", () => {
      render(<Sidebar/>, {wrapper: BrowserRouter});
      const header = screen.getByRole("banner")
      expect(header).toBeInTheDocument()
   })
   test("should render header image", () => {
      render(<Sidebar/>, {wrapper: BrowserRouter});
      const img = screen.getByRole("img")
      expect(img).toBeInTheDocument()
   })
   test("should render user first name", () => {
      render(<Sidebar/>, {wrapper: BrowserRouter});
      const firstName = screen.getByTestId("first-name")
      expect(firstName).toBeInTheDocument()
      expect(firstName).toHaveClass("name")
   })
   test("should render user last name", () => {
      render(<Sidebar/>, {wrapper: BrowserRouter});
      const lastName = screen.getByTestId("last-name")
      expect(lastName).toBeInTheDocument()
      expect(lastName).toHaveClass("name")
   })
   test("should render user role", () => {
      render(<Sidebar/>, {wrapper: BrowserRouter});
      const empRole = screen.getByTestId("role")
      expect(empRole).toBeInTheDocument()
      expect(empRole).toHaveClass("role")
   })
})

describe("nav links", () => {
   test("should render nav links", () => {
      render(<Sidebar/>, {wrapper: BrowserRouter});
      const links = screen.getAllByTestId("sidebar-link")
      expect(links.length).toBe(5)
   })

   test("should render toggle switch", () => {
      render(<Sidebar/>, {wrapper: BrowserRouter});
      const toggle = screen.getByTestId("toggle-switch")
      expect(toggle).toBeInTheDocument()
      expect(toggle).toHaveClass("toggle-switch")
   })
})