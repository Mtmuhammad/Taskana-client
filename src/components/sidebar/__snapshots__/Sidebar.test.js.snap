// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smoke and snapshot tests should match snapshot 1`] = `
<DocumentFragment>
  <nav
    class="sidebar"
  >
    <header>
      <div
        class="image-text"
      >
        <span
          class="image"
        >
          <img
            alt="Taskana alt"
            src="taskana-alt.svg"
          />
        </span>
        <div
          class="text logo-text"
        >
          <span
            class="name"
            data-testid="first-name"
          />
          <span
            class="name"
            data-testid="last-name"
          >
             
          </span>
          <span
            class="role"
            data-testid="role"
          />
        </div>
      </div>
      <i
        class="bx bx-menu toggle"
      />
    </header>
    <div
      class="menu-bar"
    >
      <div
        class="menu"
      >
        <li
          class="search-box"
        >
          <i
            class="bx bx-search icon"
          />
          <input
            disabled=""
            placeholder="Search..."
            type="text"
          />
        </li>
        <ul
          class="p-0 menu-links"
        >
          <li
            class="nav-link"
          >
            <a
              aria-current="page"
              class="sidebar-links active"
              data-testid="sidebar-link"
              href="/"
            >
              <i
                class="bx bx-home-alt icon"
              />
              <span
                class="text nav-text"
              >
                Dashboard
              </span>
            </a>
          </li>
          <li
            class="nav-link"
          >
            <a
              class="sidebar-links"
              data-testid="sidebar-link"
              href="/projects"
            >
              <i
                class="bx bx-briefcase-alt icon"
              />
              <span
                class="text nav-text"
              >
                Projects
              </span>
            </a>
          </li>
          <li
            class="nav-link"
          >
            <a
              class="sidebar-links"
              data-testid="sidebar-link"
              href="/tasks"
            >
              <i
                class="bx bx-clipboard icon"
              />
              <span
                class="text nav-text"
              >
                Tasks
              </span>
            </a>
          </li>
          <li
            class="nav-link"
          >
            <a
              class="sidebar-links"
              data-testid="sidebar-link"
              href="/tickets"
            >
              <i
                class="bx bx-note icon"
              />
              <span
                class="text nav-text"
              >
                Tickets
              </span>
            </a>
          </li>
        </ul>
      </div>
      <div
        class="bottom-content"
      >
        <li
          class=""
        >
          <a
            class="sidebar-links"
            data-testid="sidebar-link"
            href="/logout"
          >
            <i
              class="bx bx-log-out icon"
            />
            <span
              class="text nav-text"
            >
              Logout
            </span>
          </a>
        </li>
        <li
          class="mode"
        >
          <div
            class="sun-moon"
          >
            <i
              class="bx bx-moon icon moon"
            />
            <i
              class="bx bx-sun icon sun"
            />
          </div>
          <span
            class="mode-text text"
          >
            Dark Mode
          </span>
          <div
            class="toggle-switch"
            data-testid="toggle-switch"
          >
            <span
              class="switch"
            />
          </div>
        </li>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
